_posts/2018-01-26-grpc-and-deadlines.md

```diff
@@ -0,0 +1,139 @@
+---
+layout: post
+title: gRPC and Deadlines
+published: 
+permalink: blog/deadlines
+author: Gráinne Sheerin, Google SRE
+company: Google
+company-link: https://www.google.com
+---
+
+**TL;DR Always set a deadline**. This post explains why we recommend being deliberate about setting deadlines, with useful code snippets to show you how.
+
+<!--more-->
+
+When you use gRPC, the gRPC library takes care of communication, marshalling, unmarshalling, and deadline enforcement. Deadlines allow gRPC clients to specify how long they are willing to wait for an RPC to complete before the RPC is terminated with the error `DEADLINE_EXCEEDED`. By default this deadline is a very large number, dependent on the language implementation. How deadlines are specified is also language-dependent. Some language APIs work in terms of a **deadline**, a fixed point in time by which the RPC should complete. Others use a **timeout**, a duration of time after which the RPC times out.
+
+In general, when you don't set a deadline resources will be held for all in-flight requests, and all requests can potentially reach the maximum timeout. This puts the service at risk of running out of resources like memory and crashing, as well as maximizing the latency of the service.
+
+To avoid this, services should specify the longest default deadline they technically support, and clients should wait until the response is no longer useful to them. For the service this can be as simple as providing a comment in the .proto file. For the client this involves setting useful deadlines.
+
+There is no single answer to "What is a good deadline/timeout value?". Your service might be as simple as the [Greeter](https://github.com/grpc/grpc/blob/master/examples/protos/helloworld.proto) in our quick start guides, in which case 100 ms would be fine. Your service might be as complex as a globally-distributed and strongly consistent database. The deadline for a client query will be different from how long they should wait for you to drop their table.
+
+So what do you need to consider to make an informed choice of deadline? Factors to take into account include the end to end latency of the whole system, which RPCs are serial, and which can be made in parallel. You should to be able to put numbers on it, even if it's a rough calculation. Engineers need to understand the service and then set a deliberate deadline for the RPCs between clients and servers.
+
+In gRPC, both the client and server make their own independent and local determination about whether the remote procedure call (RPC) was successful. This means their conclusions may not match! An RPC that finished successfully on the server side can fail on the client side. For example, the server can send the response, but the reply can arrive at the client after their deadline has expired. The client will already have terminated with the status error `DEADLINE_EXCEEDED`.
```

Comment:
What happens in this case? 
It would be nice to propose what the client should do in this case. This leaves the server and the client in an inconsistent state. Is it then best practice to (a) include logic on the server side on requests to ensure that methods are idempotent when possible and when not possible, checking if a state change is allowed, or (b) including client side logic to check if a change has been persisted on the server side when deadlines are in use? /cc @ejona86 
At the moment, this paragraph leaves users with an open question on what to do.

Reply:
This problem isn't unique to deadlines. It can happen for I/O errors as well. Idempotency is good. But in some cases it may need the client to re-evaluate its assumptions (which may retry multiple calls in sequence).

Reply:
Could we in this case suggest adding a line to address this point. 
eg: "Clients/Servers should check for, and manage this at an application level"

@ejona86 sounds like we really do need an advanced error handling blog post. It would be useful to call out a few of these corner cases and the factors that need to be considered when handling them. :)

Reply:
That's a good idea. For final clarifications on this post can I add Grainne (who actually wrote the blog post) to this discussion as she may have some thoughts? She's @gsheerin.

Reply:
Hi (thanks all for the reviews and feedback!)

I like the suggestion "Clients and Servers should check for, and manage this at an application level". 
So that would become...
"In gRPC, both the client and server make their own independent and local determination about whether the remote procedure call (RPC) was successful. This means their conclusions may not match! An RPC that finished successfully on the server side can fail on the client side. For example, the server can send the response, but the reply can arrive at the client after their deadline has expired. The client will already have terminated with the status error `DEADLINE_EXCEEDED`.  This should be checked for and managed at the application level."


Can we also add that we'll cover that discussion, in a future blog post?



Some additional thoughts.
- If you miss a deadline (on the client side) you can retry. It has often lead to cascading failures. We need to think carefully about re-try logic before implementing, aka  "if at first you don't succeed, back off exponentially" -- Dan Sandler, Google Software Engineer.
- If your handling methods that are idempotent, it's easier. Be sensible about the re-try.
- If they aren't idempotent,  it's a decision about how to fail. It's probably better to retry deadline exceeded failures to see if it succeeded before retrying the rpc to apply the change again. 
- If you need to make multiple changes to handle a full client request, it's best to check the deadline left at each stage before attempting to preform more work, on the request on the server side. 

I'd be happy to write more material for posts, and I can help out with the error handling one if that helps.